README
QMTEST_GCC
==========

QMTEST_GCC contains emulation of the GCC DejaGNU testsuite using
QMTest (http://www.qmtest.com).

The emulation is not yet complete.  At present, the G++ testsuite, and
someof the GCC testsuite, work correctly when using a native compiler.

License
-------

QMTC is licensed under the GNU General Public License.  See the file
COPYING for the exact text of the license.

Additional Information
----------------------

For more information about this software, please contact:

  info@codesourcery.com


ChangeLog
2005-12-02  Stefan Seefeld  <stefan@codesourcery.com>

	* classes.qmc: Moved to...
	* extensions/classes.qmc: ...here.
	* compat_test.py: Moved to...
	* extensions/compat_test.py: ...here.
	* debug_test.py: Moved to...
	* extensions/debug_test.py: ...here.
	* dg_pch_test.py: Moved to...
	* extensions/dg_pch_test.py: ...here.
	* dg_tls_test.py: Moved to...
	* extensions/dg_tls_test.py: ...here.
	* gcc_database.py: Moved to...
	* extensions/gcc_database.py: ...here.
	* gcc_dg_test.py: Moved to...
	* extensions/gcc_dg_test.py: ...here.
	* gcc_dg_test_base.py: Moved to...
	* extensions/gcc_dg_test_base.py: ...here.
	* gcc_init.py: Moved to...
	* extensions/gcc_init.py: ...here.
	* gcc_test_base.py: Moved to...
	* extensions/gcc_test_base.py: ...here.
	* gcov_test.py: Moved to...
	* extensions/gcov_test.py: ...here.
	* gpp_dg_test.py: Moved to...
	* extensions/gpp_dg_test.py: ...here.
	* gpp_gcov_test.py: Moved to...
	* extensions/gpp_gcov_test.py: ...here.
	* gpp_init.py: Moved to...
	* extensions/gpp_init.py: ...here.
	* gpp_old_deja_test.py: Moved to...
	* extensions/gpp_old_deja_test.py: ...here.
	* gpp_profile_test.py: Moved to...
	* extensions/gpp_profile_test.py: ...here.
	* gpp_test_base.py: Moved to...
	* extensions/gpp_test_base.py: ...here.
	* maximal_prefix.py: Moved to...
	* extensions/maximal_prefix.py: ...here.
	* profile_test.py: Moved to...
	* extensions/profile_test.py: ...here.
	* v3_database.py: Moved to...
	* extensions/v3_database.py: ...here.
	* v3_test.py: Moved to...
	* extensions/v3_test.py: ...here.
	* setup.py: Adjust to use new installation infrastructure for
	qmtest extensions.


2005-09-06  Stefan Seefeld  <stefan@codesourcery.com>

	* gcc_dg_test_base.py: Rename __GetOutputFile to _GetOutputFile
	to make it publicly accessible.

2005-07-20  Mark Mitchell  <mark@codesourcery.com>

	* setup.py: New script.

2005-07-19  Mark Mitchell  <mark@codesourcery.com>

	* gcc_test_base.py (GCCTestBase._Compile): Use
	CompilerTable.compilers, not CompilerTable.compiler_table.
	* gpp_init.py (GPPInit.SetUp): Likewise.
	* v3_test.py (V3Init.SetUp): Likewise.
	(V3DGTest._RequireIconv): Likewise.
	
	* gcc_test_base.py (GCCTestBase._Compile): Correct typo.

2005-04-15  Stefan Seefeld  <stefan@codesourcery.com>

	* gcc_dg_test.py: Adjust filter to change in option format.
	* gcc_test_base.py: Support new compiler's ldflags.
	* gpp_init.py: Make 'GPPInit.library_directories' variable
	optional.
	* gpp_test_base.py: Likewise.
	



gpp_init.py
########################################################################
#
# File:   gpp_init.py
# Author: Mark Mitchell
# Date:   04/19/2003
#
# Contents:
#   GPPInit
#
# Copyright (c) 2003 by CodeSourcery, LLC.  All rights reserved. 
#
########################################################################

########################################################################
# Imports
########################################################################

from   compiler import CompilerExecutable
from   compiler_table import CompilerTable
from   dejagnu_base import DejaGNUBase
from   qm.executable import RedirectedExecutable
from   qm.test.resource import Resource
from   qm.test.result import Result
import os
import sys

########################################################################
# Classes
########################################################################

class GPPInit(Resource, DejaGNUBase):
    """A 'GPPInit' resource stores information for G++ tests.

    Every C++ test depends on a 'GPPInit' resource."""

    def SetUp(self, context, result):

        super(GPPInit, self)._SetUp(context)
        
        # Find the compiler table.
        compilers = context["CompilerTable.compilers"]

        # Get the C++ compiler.
        compiler = compilers["cplusplus"]

        # Tell GCOVTest where the coverage executable is located.
        gcov = os.path.join(os.path.dirname(compiler.GetPath()),
                            "gcov")
        context["GCOVTest.gcov"] = gcov
        
        # Run the compiler to find out what multilib directory is
        # in use. The DejaGNU code that does this is get_multiblis in
        # libgloss.exp; this version uses a much simpler technique.
        options = compiler.GetOptions()
        executable = CompilerExecutable()
        executable.Run([compiler.GetPath()]
                       + options
                       + ['--print-multi-dir'])
        directory = executable.stdout[:-1]

        # Assume that no additional library directories need to be
        # explicitly provided to the compiler. 
        context["GPPInit.library_directories"] = []
        
        # If the compiler is being run out of the build directory,
        # perform special set up associated with that configuration.
        # To determine whether the compiler is being run out of the
        # build directory, we find the compiler by looking for the
        # first -B option.  (This is the technique used by
        # get_multilibs in libgloss.exp.)
        for o in options:
            if o.startswith("-B"):
                objdir = os.path.dirname(os.path.dirname(o[2:]))
                options.append(self.__SetUpInObjdir(result, context, objdir))
                break
        

        # Avoid splitting diagnostic message lines.
        options.append("-fmessage-length=0")
        # Remember the options to use.
        context["GPPInit.options"] = options
        

    def __SetUpInObjdir(self, result, context, objdir):
        """Setup for a compiler being run out of the build directory.

        'objdir' -- The path to the 'objdir' directory, i.e., the
        directory that contains 'xgcc'.
        
        'result' -- As for 'SetUp'.

        'context' -- As for 'SetUp'.

        returns -- A list of additional command-line options that
        should be provided to the compiler.
        
        It the compiler being tested is not an installed compiler, but
        is in in the GCC build directory, then additional command-line
        options must be passed to 'g++' to make it aware of the
        location of various components."""

        # Compute the path to the V3 object directory.  See 'g++_init'
        # in the GCC testsuite for the DejaGNU equivalent to this
        # code.
        target = context["DejaGNUTest.target"]
        v3_directory = os.path.normpath(os.path.join(objdir,
                                                     target,
                                                     directory,
                                                     "libstdc++-v3"))

        # Run "testsuite_flags" to figure out which -I options to use
        # when running tests.
        command = [os.path.join(v3_directory,
                                "scripts",
                                "testsuite_flags"),
                   "--build-includes"]
        result["GPPInit.testsuite_flags_command"] \
            = result.Quote(" ".join(command))
        try:
            executable = RedirectedExecutable()
            executable.Run(command)
            options = executable.stdout.split()
        except:
            result.NoteException(cause="Could not run testsuite_flags",
                                 outcome=Result.FAIL)
            return []

        # Avoid splitting diagnostic message lines.
        options.append("-fmessage-length=0")
        # Remember the options to use.
        context["GPPInit.options"] = options

        # Compute the directories containing runtime libraries.  These
        # directories must be provided both at link-time (in the form
        # of -L options) and at run-time (in LD_LIBRARY_PATH).
        lib_dirs = []
        lib_dirs.append(os.path.join(v3_directory, "src", ".libs"))
        lib_dirs.append(os.path.join(objdir, "gcc"))
        context["GPPInit.library_directories"] = lib_dirs
    
        return options
